### DISCUSSION TITLE
paddlespeech在windows上实现cpu的训练

### DISCUSSION THREAD

**Discussion by Doubledongli** [PaddlePaddle/PaddleSpeech#2156] [Ideas]
> 各位，windows上面使用ppspeech上进行训练的时候，你一定报过很多错，主要是warp_ctc 那块的Cpython目前不支持win上面训练，为了解决这个问题，我对代码做了一些调整，可以参考，现在只是支持cpu，gpu还不支持。
### 1、修改warp_ctc部分的代码，用其他模块实现
**修改位置**：
D:\conda\envs\asr\Lib\site-packages\paddle\fluid\layers\loss.py
第600行

```python
        # grad, loss_out = _C_ops.warpctc(
        #     input,
        #     label,
        #     input_length,
        #     label_length,
        #     'blank',
        #     blank,
        #     'norm_by_times',
        #     norm_by_times, )
        # using Tensor
        import paddle
        import paddle.fluid as fluid
        import numpy as np
        # 注意：下面的几个数值，181、151、8和31是我debug源代码的时候改的，这个地方最好改成解包后动判断，这部分抽空实现下
        # length of the longest logit sequence
        max_seq_length = 181
        # length of the longest label sequence
        max_label_length = 151
        # number of logit sequences
        batch_size = 8
        # class num
        class_num = 31
        paddle.enable_static()
        logits = fluid.data(name='logits',
                            shape=[max_seq_length, batch_size, class_num + 1],
                            dtype='float32')
        logits_length = fluid.data(name='logits_length', shape=[None],
                                   dtype='int64')
        label = fluid.data(name='label', shape=[batch_size, max_label_length],
                           dtype='int32')
        label_length = fluid.data(name='labels_length', shape=[None],
                                  dtype='int64')
        cost = fluid.layers.warpctc(input=logits, label=label,
                                    input_length=logits_length,
                                    label_length=label_length)
        place = fluid.CPUPlace()
        x = np.random.rand(max_seq_length, batch_size, class_num + 1).astype("float32")
        y = np.random.randint(0, class_num, [batch_size, max_label_length]).astype("int32")
        exe = fluid.Executor(place)
        output = exe.run(fluid.default_main_program(),
                         feed={"logits": x,
                               "label": y,
                               "logits_length": np.array([max_seq_length] * batch_size).astype("int64"),
                               "labels_length": np.array([max_label_length] * batch_size).astype("int64")},
                         fetch_list=[cost.name])
        paddle.disable_static()
        loss_out = paddle.to_tensor(output[0])
 ```

如上，注释掉原来的代码，改成下面的实现方式，原来不可以训练的就可以实现如下的训练了：
![image](https://user-images.githubusercontent.com/20540661/179177699-f4cd42e4-6ea2-46c3-afe9-d309a5bb4f6f.png)
### 2、之前的报错
![1656984090417_673C5E13-C3DB-423c-865F-93A0E8BF2F07](https://user-images.githubusercontent.com/20540661/179178823-08ef8e85-32f4-4fcf-831a-d38d5bf981be.png)
### 3、我的pip版本列表
```bash
pip list
```

```text
Package                       Version
----------------------------- ------------
aiohttp                       3.8.1
aiosignal                     1.2.0
anyio                         3.6.1
appdirs                       1.4.4
astor                         0.8.1
async-timeout                 4.0.2
asyncio                       3.4.3
asynctest                     0.13.0
attrs                         21.4.0
audioread                     2.1.9
Babel                         2.10.3
backcall                      0.2.0
backports.functools-lru-cache 1.6.4
bce-python-sdk                0.8.73
bleach                        5.0.1
Bottleneck                    1.3.4
build                         0.8.0
certifi                       2022.6.15
cffi                          1.15.0
cfgv                          3.3.1
charset-normalizer            2.0.12
click                         8.1.3
colorama                      0.4.5
colorlog                      6.6.0
commonmark                    0.9.1
cycler                        0.11.0
Cython                        0.29.30
datasets                      2.3.2
debugpy                       1.6.0
decorator                     4.4.2
dill                          0.3.4
Distance                      0.1.3
distlib                       0.3.4
docutils                      0.19
editdistance                  0.6.0
entrypoints                   0.4
fastapi                       0.78.0
filelock                      3.7.1
flake8                        4.0.1
Flask                         2.1.2
Flask-Babel                   2.0.0
flatbuffers                   2.0
fonttools                     4.33.3
frozenlist                    1.3.0
fsspec                        2022.5.0
future                        0.18.2
g2p-en                        2.1.0
g2pM                          0.1.2.5
gast                          0.3.3
h11                           0.13.0
h5py                          3.7.0
huggingface-hub               0.8.1
identify                      2.5.1
idna                          3.3
importlib-metadata            4.2.0
inflect                       5.6.0
ipykernel                     6.15.0
ipython                       7.33.0
itsdangerous                  2.1.2
jedi                          0.18.1
jieba                         0.42.1
Jinja2                        3.1.2
joblib                        1.1.0
jsonlines                     3.0.0
jupyter-client                7.3.4
jupyter-core                  4.10.0
kaldiio                       2.17.2
keyring                       23.6.0
kiwisolver                    1.4.3
librosa                       0.8.1
llvmlite                      0.38.1
loguru                        0.6.0
MarkupSafe                    2.1.1
matplotlib                    3.5.2
matplotlib-inline             0.1.3
mccabe                        0.6.1
mock                          4.0.3
multidict                     6.0.2
multiprocess                  0.70.12.2
nara-wpe                      0.0.8
nest-asyncio                  1.5.5
nltk                          3.7
nodeenv                       1.7.0
numba                         0.55.2
numpy                         1.19.3
onnxruntime                   1.11.1
opencv-python                 4.6.0.66
opt-einsum                    3.3.0
packaging                     21.3
paddle-bfloat                 0.1.7
paddle2onnx                   0.9.8
paddleaudio                   1.0.1
paddlefsl                     1.1.0
paddlenlp                     2.3.1
paddlepaddle                  2.3.1
paddlespeech                  1.0.1
paddlespeech-ctcdecoders      0.2.0
paddlespeech-feat             0.1.0
pandas                        1.1.5
parso                         0.8.3
pathos                        0.2.8
pattern-singleton             1.2.0
pep517                        0.12.0
pickleshare                   0.7.5
Pillow                        9.1.1
pip                           22.1.2
pkginfo                       1.8.3
platformdirs                  2.5.2
pooch                         1.6.0
portalocker                   2.4.0
pox                           0.3.1
ppft                          1.7.6.5
praatio                       5.0.0
pre-commit                    2.19.0
prettytable                   3.3.0
prompt-toolkit                3.0.30
protobuf                      3.20.0
psutil                        5.9.1
pyarrow                       8.0.0
PyAudio                       0.2.11
pycodestyle                   2.8.0
pycparser                     2.21
pycryptodome                  3.15.0
pydantic                      1.9.1
pyflakes                      2.4.0
Pygments                      2.12.0
pyparsing                     3.0.9
pypinyin                      0.46.0
pypinyin-dict                 0.2.0
pytest-runner                 6.0.0
python-dateutil               2.8.2
pytz                          2022.1
pywin32                       304
pywin32-ctypes                0.2.0
pyworld                       0.3.0
PyYAML                        6.0
pyzmq                         23.2.0
readme-renderer               35.0
regex                         2022.6.2
requests                      2.28.0
requests-toolbelt             0.9.1
resampy                       0.2.2
responses                     0.18.0
rfc3986                       2.0.0
rich                          12.5.1
sacrebleu                     2.1.0
scikit-learn                  1.0.2
scipy                         1.7.3
sentencepiece                 0.1.96
seqeval                       1.2.2
setuptools                    62.6.0
shellcheck-py                 0.8.0.4
six                           1.16.0
sniffio                       1.2.0
SoundFile                     0.10.3.post1
starlette                     0.19.1
tabulate                      0.8.10
TextGrid                      1.5
threadpoolctl                 3.1.0
timer                         0.2.2
toml                          0.10.2
tomli                         2.0.1
torch                         1.12.0
tornado                       6.2
tqdm                          4.64.0
traitlets                     5.3.0
twine                         4.0.1
typeguard                     2.13.3
typing_extensions             4.2.0
urllib3                       1.26.9
uvicorn                       0.18.2
virtualenv                    20.15.0
visualdl                      2.2.3
wcwidth                       0.2.5
webencodings                  0.5.1
webrtcvad                     2.0.10
websockets                    10.3
Werkzeug                      2.1.2
wheel                         0.37.1
win32-setctime                1.1.0
xxhash                        3.0.0
yacs                          0.1.8
yarl                          1.7.2
zhon                          1.1.5
zipp                          3.8.0
```
